### DISCUSSION TITLE
How to write a client library in Crystal?

### DISCUSSION THREAD

**Discussion by notramo** [geldata/gel#4295] [General]
I am curious how much work would it be to write a query-only, unencrypted (localhost), minimal client library.
I guess it depends on if it's possible to generate data types and code for serialization somehow, or it's only possible to manually write every data type? I'm interested in writing it in [Crystal](https://crystal-lang.org/). It's a compiled, statically typed language, with uniqe and innovative type system (automatic type unions combined with flow typing). It nearly feels like a dynamically typed language (flexibility), but eliminates runtime type errors.
I'm not an experienced programmer, but I would at least try it if it's not too difficult.

**Comment by 1st1** [comment]
Really hard to answer. Making a PoC kinda library can take anywhere from 1 week to 2. And TLS/non-TLS shouldn't make much difference, the difference will be in just a couple of APIs calls.

**Comment by notramo** [comment]
Is it possible to generate code for protocol data structures from some protocol spec?

**Reply by fmoor** [reply]
> Is it possible to generate code for protocol data structures from some protocol spec?

I don't think there is anything that would make this convenient. The only things you could possibly parse is the documentation or existing client implementations.

**Comment by colinhacks** [comment]
@notramo If this is for development primarily, I'd recommend implementing a wrapper over the HTTP interface: https://www.edgedb.com/docs/clients/http/index

It has lots of downsides over a proper client (slower, no rich datatypes, no support for transactions, etc) but it'll get you up and running.

**Comment by zetashift** [comment]
Is there any tips/tutorials on how to write a driver?

**Reply by fmoor** [reply]
There aren't any tutorials that I am aware of. The documentation for the binary protocol is [here](https://www.edgedb.com/docs/reference/protocol/index). I started by running simple queries with the python driver and deciphering the bytes that were sent to/from the server. Then adding the same behavior to my own driver.

**Comment by jaclarke** [comment]
Did a little playing around with Crystal, and just in case it's useful to anyone that wants to write a Crystal client, here's a basic example of how to initalise a connection to edgedb over the binary protocol: https://gist.github.com/jaclarke/1c73e90454aa50f3df7d98af3554d93f

Note: I had to fix an alpn bug in Crystal's stdlib, so for now if you want to try this out, you'll need to clone this branch: https://github.com/jaclarke/crystal/tree/client_alpn, and use `bin/crystal` from there (see: https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md#the-standard-library)

**Reply by 0x241F31** [reply]
@jaclarke was this bug in Crystal stdlib fixed?

**Reply by jaclarke** [reply]
Yes, it was fixed in Crystal 1.7.0
